keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/DoubleVertex.java

```diff
@@ -1,193 +1,221 @@
 package io.improbable.keanu.vertices.dbl;
 
 
+import java.util.Map;
+import java.util.function.Function;
+
 import io.improbable.keanu.kotlin.DoubleOperators;
+import io.improbable.keanu.tensor.NumberTensor;
+import io.improbable.keanu.tensor.Tensor;
 import io.improbable.keanu.tensor.dbl.DoubleTensor;
-import io.improbable.keanu.vertices.ContinuousVertex;
+import io.improbable.keanu.vertices.IVertex;
+import io.improbable.keanu.vertices.Observable;
 import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.bool.BooleanVertex;
+import io.improbable.keanu.vertices.bool.nonprobabilistic.operators.binary.BooleanBinaryOpVertex;
 import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
 import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
-import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.binary.*;
-import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary.*;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.binary.DoubleBinaryOpVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.binary.MatrixMultiplicationVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary.DoubleUnaryOpLambda;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary.DoubleUnaryOpVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.Differentiable;
+import io.improbable.keanu.vertices.update.ValueUpdater;
 
-import java.util.*;
-import java.util.function.Function;
+public abstract class DoubleVertex extends Vertex<DoubleTensor> implements DoubleOperators<DoubleVertex>, Differentiable {
 
-public abstract class DoubleVertex extends ContinuousVertex<DoubleTensor> implements DoubleOperators<DoubleVertex> {
+    public DoubleVertex(ValueUpdater<DoubleTensor> valueUpdater, Observable<DoubleTensor> observation) {
+        super(valueUpdater, observation);
+    }
 
+    @Override
     public DoubleVertex minus(DoubleVertex that) {
-        return new DifferenceVertex(this, that);
+        return new DoubleBinaryOpVertex(this, that,  (a, b) -> a.minus(b), (a, b) -> a.minus(b));
```

Comment:
Making all ops lambdas turns the vertex into a black box. This is nice and DRY at the cost of increasing code complexity. How can we serialize these lambda vertices to json or protobuf?

Reply:
I'm not opposed to reinstating the concrete child classes again. The important thing for me is that they are now small and consistent (i.e. each concrete class will just need a constructor that calls `super`).